### PR TITLE
Optimize ROCM Prefill and Decode

### PR DESCRIPTION
**Pull request by gaetan-puleo**
## Summary

This PR adds guarded ROCm/HIP inference optimizations for RDNA3.5 and Strix Halo systems, including the Ryzen AI Max+ 395 with Radeon 8060S graphics.

Main improvements:

- RDNA3.5 MMQ/MMVQ configurations and register prefetching
- Compact `MUL_MAT_ID` execution with quant- and geometry-specific tile selection
- Fused activation quantization for Q8_0 and Q6_K decode
- Optimized MoE routing, weighted expert reduction, and shared-expert operations
- DPP and tiled Gated DeltaNet kernels for prefill and decode
- Fused hyper-connection stream mixing and combination
- D=256 WMMA Flash Attention for deep-context prefill
- Device-resident speculative decoding checkpoints
- Qwen4exp QSA and PLE graph optimizations

Architecture, shape, layout, aliasing, and graph-closure guards preserve existing behavior on unsupported devices and tensor configurations.

## Correctness

The complete PR was reviewed by five independent agents, followed by a separate source-level verification of every reported finding.

Confirmed issues were fixed, including:

- Partial routed-MMQ tile bounds
- Uninitialized MMVQ prefetch reads
- All-zero fused SwiGLU quantization
- Long-context GDN index overflow
- Fusion closure and external-consumer handling
- Deterministic MoE ID and weight ordering

Final validation:

- **14,050/14,050 ROCm backend tests passed**
- **61/61 models and quants completed model-level validation**
- **61/61 tokenizations matched latest master**
- **50/61 greedy generations matched latest master exactly**
- The remaining 11 were independently reviewed and classified as valid floating-point accumulation sensitivity, with no incorrect quantization, routing, state handling, or graph semantics found
- Qwen3.5 and Qwen3.8 CPU graph spot checks were byte-identical to master
- Qwen3.8 controlled one-token GPU logits were also byte-identical

## Performance

Representative measured improvements include:

- Qwen3.5 D=256 prefill at depth 32k: approximately **+28%**
- Qwen3.5 D=256 prefill at depth 64k: approximately **+37%**
- Qwen3.6 fused GDN decode: approximately **+9-11%**
- Qwen3.6 tiled GDN prefill: approximately **+8-11%**
- Selected routed Q4_K/Q5_K/Q6_K kernels: approximately **20-54% lower kernel time**
- Ling PP2048: approximately **+7-11%**, depending on quant
- Qwen3.5-0.8B Q8_K_XL versus latest master:
  - PP2048: `8704.46 -> 9561.11` tok/s, **+9.8%**
  - TG128: `130.67 -> 131.85` tok/s, **+0.9%**

## Benchmark Coverage

Benchmarked on gfx1151 with ROCm 7.14:

- **61 models**
- **1,525 configurations**
- PP512, PP1024, PP2048, PP4096, and TG128
- Depths 0, 12k, 32k, 64k, and 128k
- Dense models: `-ub 512`
- MoE models: `-ub 2048`
- Common options: `-b 2048 -fa 1 -ngl 999 --load-mode mmap -r 4`
- First repetition discarded; results use repetitions 2-4

Interactive and raw results:

https://gaetan-puleo.github.io/strix-halo-pull-18-results/

The branch includes latest master `b212548e0`.

Final head: `756f0d443`

### CODE REVIEW

**Comment by dzannotti** [comment]
Thanks for putting this together — the RDNA3.5 work here is nice, and the `551ce30fe` follow-up (keeping batch-one MMVQ on one wave to preserve exact FP32 reduction order) is a good catch and exactly the kind of trade we'd want made that way.

We had a Strix Halo box free tonight, so we ran an independent A/B to help fill in the Measurements section. Sharing the raw numbers in case they're useful — happy to re-run anything with different settings.

## Setup

```
Device:     Ryzen AI Max+ 395 / Radeon 8060S (gfx1151)
Memory:     128 GB LPDDR5X, UMA; 120832 MiB reported addressable
Kernel:     7.1.8-200.fc44.x86_64
Backend:    ROCm 10.0.0 (container), AMDGPU_TARGETS=gfx1151
Build:      Release, GGML_HIP=ON, GGML_HIP_NO_VMM=ON
Baseline:   732484c20 (halo master, this PR's merge-base)
Also ran:   6c84c7d5d (upstream base), 3466812d1 (upstream master today)
Change:     551ce30fe
Models:     Qwen3.8-27B-Q4_K_M; Ornith-1.5-35B-A3B-Q4_K_M
Settings:   -fa on -lm none -b 2048 -ub 2048 -r 3 -ngl 99
```

All four trees were built and run in the same session, back to back, with nothing else on the GPU.

## Prompt processing (tok/s, ± stddev over 3 reps)

| model | test | upstream base | upstream today | halo master | **this PR** | vs halo | vs upstream |
|---|---|---:|---:|---:|---:|---:|---:|
| Qwen3.8-27B Q4_K_M | pp9000 | 329.93±0.29 | 329.81±0.16 | 331.33±0.36 | **396.70±0.73** | 1.197x | 1.203x |
| Qwen3.8-27B Q4_K_M | pp16000 | 310.71±0.04 | 310.84±0.14 | 311.38±0.03 | **371.41±0.09** | 1.193x | 1.195x |
| Ornith-1.5-35B-A3B Q4_K_M | pp9000 | 1268.80±2.93 | 1239.85±1.69 | 1263.92±6.64 | **1793.15±3.65** | 1.419x | 1.446x |
| Ornith-1.5-35B-A3B Q4_K_M | pp16000 | 1176.65±1.44 | 1153.24±2.50 | 1175.74±1.35 | **1629.05±4.94** | 1.386x | 1.413x |

## Token generation (tok/s)

| model | test | halo master | this PR | ratio |
|---|---|---:|---:|---:|
| Qwen3.8-27B Q4_K_M | tg128@9000 | 10.78±0.07 | 10.83±0.02 | 1.004x |
| Qwen3.8-27B Q4_K_M | tg128@16000 | 10.54±0.09 | 10.61±0.02 | 1.006x |
| Ornith-1.5-35B-A3B Q4_K_M | tg128@9000 | 60.55±0.49 | 60.77±0.46 | 1.004x |
| Ornith-1.5-35B-A3B Q4_K_M | tg128@16000 | 58.27±0.48 | 58.55±0.45 | 1.005x |

## Correctness

`test-backend-ops -b ROCm0` on `551ce30fe`: **13478/13478 tests passed**, Backend ROCm0: OK.

We also read through the FA `common_mask` reuse, since sharing one mask row across grouped heads looked like the highest-risk change. It looks correct to us: `slope` is pinned to `1.0f` whenever `ncols2 != 1`, so there's no ALiBi-slope hazard, and with `ncols=32`/`ncols2=8` each warp's `cpw` columns stay inside a single token group. Noting it mainly so the reasoning is written down somewhere.

## Observations

- The prefill win is real and very reproducible — stddev is under 0.7% on every cell.
- The MoE model benefits about twice as much as the dense one (~1.40x vs ~1.20x), which matches where the diff is concentrated: routed-expert kernels, `mul_mat_id` tile selection, expert aggregation. Might be worth leading with an MoE model in the PR description, since that's where the change shines.
- On ROCm the gains are flat between 9k and 16k rather than growing with depth. The baseline doesn't collapse at depth here (27B only loses ~6% from 9k to 16k), so there isn't much depth-scaling headroom to recover — unlike the pattern reported on the Vulkan side in #17.
- Decode is 1.00x across the board, which we read as the expected consequence of `551ce30fe`. Worth stating explicitly in the description so nobody expects a tg change.
- halo master measures within 1.00x of upstream on all eight cells for these two models, so on ROCm essentially all of the improvement is coming from this PR.

## Two small questions

Both look unrelated to prefill and we didn't exercise them, so flagging rather than asserting anything:

1. `common/common.h` drops `COMMON_SPECULATIVE_TYPE_DRAFT_MTP` from `need_n_rs_seq()`. Is that intended to pair with the `LLAMA_STATE_SEQ_FLAGS_ON_DEVICE` change in `server-context.cpp`?
2. `GGML_OP_CONCAT` is added to `ggml_backend_op_alloc_size_may_expand`. Is the padded-allocation invariant guaranteed for every backend that can see a quantized CONCAT, or only the HIP path?

What we did **not** verify: speculative/MTP decode paths, the server changes, Vulkan, any non-gfx1151 device, and perplexity (we ran op-level correctness only, not end-to-end output quality).

*Disclosure: these benchmarks were run and this comment was written by Claude Code, on hardware and at the request of @dzannotti. Numbers are machine-generated but reproducible with the command line above; happy to share the raw JSON.*

**Comment by gaetan-puleo** [comment]
> common/common.h drops COMMON_SPECULATIVE_TYPE_DRAFT_MTP from need_n_rs_seq(). Is that intended to pair with the LLAMA_STATE_SEQ_FLAGS_ON_DEVICE change in server-context.cpp?

This was a fix for correctness on MTP

Nathan uses it in the Vulkan Performance toolbox too

**Comment by dzannotti** [comment]
ship this! this is awesome

**Review by dzannotti — APPROVED** [review]
